### DISCUSSION TITLE
[Feature suggestion] Add rumble support for Generic Controllers

### DISCUSSION THREAD

**Discussion by JPZV** [darthcloud/BlueRetro#611] [Ideas]
I'm making a self-made BT controller with rumble support through HID (with its respective descriptors), but it don't work on BlueRetro because, after looking at the code, the rumble feature is disabled on Generic Controllers according to [bluetooth/hidp/hidp.c#L74](https://github.com/darthcloud/BlueRetro/blob/master/main/bluetooth/hidp/hidp.c#L74)

My suggestion is to, if the controller is a generic/unknown one, check its descriptors and look if it supports the `09 97` and `09 98` whose tells if it supports enabling/disabling DC actuators. Then, every time its needed, send the rumble data to the report_id for those descriptors. This would not only help with generic controllers like mine, but also should work with controllers like the [Stadia one](https://github.com/darthcloud/BlueRetro/issues/578)

Just in case, here is my controller HID Report parsed with RDD! HID Report Descriptor Decoder, specific on the DC actuators section:

<details>
  <summary>HID Report</summary>
  
```
05 01        (GLOBAL) USAGE_PAGE         0x0001 Generic Desktop Page
09 05        (LOCAL)  USAGE              0x00010005 Game Pad (Application Collection)
A1 01        (MAIN)   COLLECTION         0x01 Application (Usage=0x00010005: Page=Generic Desktop Page, Usage=Game Pad, Type=Application Collection)
...
05 0F          (GLOBAL) USAGE_PAGE         0x000F Physical Interface Device Page
09 95          (LOCAL)  USAGE              0x000F0095 PID Device Control Report (Logical Collection)
A1 02          (MAIN)   COLLECTION         0x02 Logical (Usage=0x000F0095: Page=Physical Interface Device Page, Usage=PID Device Control Report, Type=Logical Collection)
85 04            (GLOBAL) REPORT_ID          0x04 (4)
09 96            (LOCAL)  USAGE              0x000F0096 PID Device Control (Named Array Collection)
A1 02            (MAIN)   COLLECTION         0x02 Logical (Usage=0x000F0096: Page=Physical Interface Device Page, Usage=PID Device Control, Type=Named Array Collection)
09 97              (LOCAL)  USAGE              0x000F0097 DC Enable Actuators (Selector)
09 98              (LOCAL)  USAGE              0x000F0098 DC Disable Actuators (Selector)
09 99              (LOCAL)  USAGE              0x000F0099 DC Stop All Effects (Selector)
09 9A              (LOCAL)  USAGE              0x000F009A DC Device Reset (Selector)
09 9B              (LOCAL)  USAGE              0x000F009B DC Device Pause (Selector)
09 9C              (LOCAL)  USAGE              0x000F009C DC Device Continue (Selector)
15 01              (GLOBAL) LOGICAL_MINIMUM    0x01 (1)
25 06              (GLOBAL) LOGICAL_MAXIMUM    0x06 (6)
75 08              (GLOBAL) REPORT_SIZE        0x08 (8) Number of bits per field
95 01              (GLOBAL) REPORT_COUNT       0x01 (1) Number of fields
91 02              (MAIN)   OUTPUT             0x00000002 (1 field x 8 bits) 0=Data 1=Variable 0=Absolute 0=NoWrap 0=Linear 0=PrefState 0=NoNull 0=NonVolatile 0=Bitmap
C0               (MAIN)   END_COLLECTION     Logical
C0             (MAIN)   END_COLLECTION     Logical
C0           (MAIN)   END_COLLECTION     Application
```
</details>

**Comment by DJm00n** [comment]
Actually this can be used for Xbox One and Luna controllers too.
So result we can try to do something generic code that should work for all Xbox/Stadia/Luna and @JPZV's controllers instead of adding vibration support one-by-one model:
```
if (hid_desc_have (Usage 0x000F0097)) // DC Enable Actuators

 set(Usage 0x000F0097, vibration_strength mapped to usage min/max) // DC Enable Actuators
 set(Usage 0x000F0070, vibration_strength mapped to usage min/max) // Magnitude
 set(Usage 0x000F0050, maximum) // Duration
 set(Usage 0x000F00A7, minimum) // Start Delay
 set(Usage 0x000F007C, maximum) // Loop Count

```

**Comment by darthcloud** [comment]
My focus now is to make Bluetooth connection more reliable and as such anything rumble related is a the bottom of my list.

**Reply by JPZV** [reply]
Would be there a problem if I or someone else tries to implement it?

**Reply by darthcloud** [reply]
its not a problem but PR review are still time consuming, so it may stay pending for a while until I get to it.
